docs/proposals/predictive-horizontal-pod-autoscaling.md

```diff
@@ -0,0 +1,232 @@
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Predictive Horizontal Pod Autoscaling
+
+**Author**: Matt McNaughton (@mattjmcnaughton)
+
+## Preface
+
+This document describes the addition of prediction to the
+[horizontal pod autoscaler] (../user-guide/horizontal-pod-autoscaler.md).
+This addition will decrease the amount of time pods spend operating outside of
+their target value for resource consumption.
+
+## Overview and Motivation
+
+In 1.1, Kubernetes added the experimental horizontal autoscaling of pods,
+ensuring the number of pods automatically adjusts as applications’ demands rise
+and fall. This autoscaling, described in detail in the [documentation]
+(../user-guide/horizontal-pod-autoscaler.md), works based on the following
+calculation:
+
+```
+TargetNumOfPods = ceil(sum(CurrentPodsResourceConsumption) / TargetResourceConsumption)
+```
+
+If the selected pods are operating outside of their target resource
+consumption, the autoscaler will destroy or create pods until all the selected
+pods are operating in accordance with target consumption.
+
+Overall, this method of autoscaling works well, yet it could
+be improved by accounting for the amount of time necessary
+to initialize a pod (i.e. for a pod to be ready to share the computational
+work). A pod’s initialization time can be lengthy if the pod must initialize
+a framework, download shard files from a
+network, register itself with some type of manager, etc.
+More specifically, if the newly created pods take a long time to
+initialize, the already existing pods will spend the entire
+initialization time operating outside of target
+resource consumption while they wait for the newly created pods to
+share the computational work. Depending on pod initialization time, and the
+danger of operating outside target resource consumption, it can be
+extremely detrimental to not account for pod initialization time.
+
+**Predictive horizontal pod autoscaling** addresses this problem.
+It accounts for pod initialization time by
+determining the target number of pods based on the predicted future pod resource
+consumption, instead of the current pod resource consumption. In other words,
+if it takes 10 minutes for a pod to initialize, than if it is 6:00pm and
+predictive autoscaling is enabled, autoscaling behavior is determined by a
+prediction of the application’s demands at 6:10pm. If it is determined that a
+new instance of this pod is needed based on the prediction, the new pod will be
+created at 6:00pm, meaning it will be initialized by 6:10pm. The pods will
+never have to operate outside of their target resource consumption (as opposed
+to if autoscaling was occurring in a non-predictive manner, in which case the
+pod would not be created until 6:10 and would not be initialized until 6:20pm,
+meaning there was 10 minutes in which pods were operating outside of target
```

Comment:
Right now HPA uses 1 minute of data to predict what the use will be in near future. To reliably guess (and don't waste 10 min for nothing) more about the future you will need more than 1 minute of data.